File: cs/UCLA/libUCLA/UServer.cs
using System;

using CrossroadsIO;

namespace libUCLA {
    public delegate void UReceiveHandler(object sender, UDataReceivedArgs e);

    public class UDataReceivedArgs : EventArgs {
        public byte[] Buffer { get; set; }

        public UDataReceivedArgs(byte[] buffer): base() {
            this.Buffer = buffer;
        }
    }

    public class UServer : IDisposable {
        public const int MaxDataLength = 256;

        /// <summary>
        /// True if server has been started
        /// </summary>
        public bool IsStarted {
            get { return this.isStarted; }
        }

        /// <summary>
        /// This handler is called every time when message is received.
        /// </summary>
        public event UReceiveHandler DataReceived;

        private bool isStarted = false;
        private Context ctx;
        private Socket sock;
        private string endpoint;

        private bool _disposed;

        /// <summary>
        /// UCLA Server constructor.
        /// Creates server instance from specified config.
        /// </summary>
        /// <param name="config">Reference to configuration object.</param>
        /// <param name="autostart">Tells if server will be started automatically.</param>
        public UServer(UConfig config, bool autostart = false) {
            this.endpoint = config.Endpoint;

            if (autostart) {
                this.Start();
            }
        }

        /// <summary>
        /// Starts a server.
        /// </summary>
        public void Start() {
            EnsureNotDisposed();

            if (this.isStarted) return;

            this.ctx = Context.Create();
            this.sock = this.ctx.CreateSocket(SocketType.PULL);

            this.sock.Bind(this.endpoint);

            this.isStarted = true;
        }

        /// <summary>
        /// Receives single message.
        /// </summary>
        public void Receive() {
            EnsureNotDisposed();

            byte[] buf = new byte[UServer.MaxDataLength];
            int receivedLength = this.sock.Receive(buf);

            if (this.DataReceived != null) {
                byte[] callBuf = new byte[receivedLength];
                Array.Copy(buf, callBuf, receivedLength);

                this.DataReceived(this, new UDataReceivedArgs(callBuf));
            }
        }

        /// <summary>
        /// Run in loop and receive messages.
        /// NOTE: This is infinite loop - just don't expect it to stop.
        /// </summary>
        public void Run() {
            EnsureNotDisposed();

            while (true) {
                this.Receive();
            }
        }

        public void Dispose() {
            this.Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing) {
            if (!this._disposed) {
                if (disposing) {
                    this.sock.Close();
                    this.ctx.Terminate();

                    this.sock.Dispose();
                    this.ctx.Dispose();
                }

                this._disposed = true;
            }
        }

        private void EnsureNotDisposed() {
            if (this._disposed) {
                throw new ObjectDisposedException(GetType().FullName);
            }
        }

    }
}

File: cpp/UCLA/libUCLA/UServer.cpp
#include "UServer.h"
#include "config.h"

#ifdef UCLA_HAVE_UNIX
#include <xs/xs.hpp>
#else
#include <xs.hpp>
#endif

using namespace UCLA;

UServer::UServer(UConfig &config, bool autostart){
	this->_isStarted = false;
	this->_receive_handler = NULL;
	this->_ctx = NULL;
	this->_sock = NULL;
	this->_endpoint = NULL;

	CopyEndpoint(config.Endpoint());

	if(autostart){
		Start();
	}
}

void UServer::Start(void){
	if(this->_isStarted){
		return;
	}

	try{
		this->_ctx = new xs::context_t();
		this->_sock = new xs::socket_t(*(static_cast<xs::context_t*>(this->_ctx)), XS_PULL);

		static_cast<xs::socket_t*>(this->_sock)->bind(this->_endpoint);
	}catch(xs::error_t &err){
		throw UException(err.what());
	}

	this->_isStarted = true;
}

void UServer::Receive(void){
	if(!this->_isStarted){
		this->Start();
	}

	char *buf = new char[UCLA_MAX_MESSAGE_LEN];
	int received_len = -1;

	try{
		received_len = static_cast<xs::socket_t*>(this->_sock)->recv(buf, UCLA_MAX_MESSAGE_LEN);
	}catch(xs::error_t &err){
		throw UException(err.what());
	}

	if(this->_receive_handler != NULL){
		this->_receive_handler(buf, received_len);
	}

	delete[] buf;
}

void UServer::Run(void){
	while(true){
		this->Receive();
	}
}

void UServer::SetupReceiveHandler(UCLA_RECEIVE_HANDLER handler){
	this->_receive_handler = handler;
}

UServer::UServer(UServer &&that){
	if(this->_isStarted)
	{
		this->CleanUpXS();
	}

	this->_isStarted = false;
	this->_ctx = NULL;
	this->_sock = NULL;
	this->_endpoint = NULL;

	CopyEndpoint(that.Endpoint());

	that.~UServer();
}

UServer& UServer::operator=(UServer&& that){
	if (this != &that){
		if(this->_isStarted)
		{
			this->CleanUpXS();
		}

		this->_isStarted = false;
		this->_ctx = NULL;
		this->_sock = NULL;

		CopyEndpoint(that.Endpoint());

		that.~UServer();
	}

	return *this;
}

void UServer::CopyEndpoint(const char* endpoint){
	using namespace Utils;

	if(this->_endpoint != NULL)
		delete[] this->_endpoint;

	int strLength = strlen(endpoint) + 1;

	char* tmp_endpoint = new char[strLength];
	this->_endpoint = tmp_endpoint;

	m_strlcpy(this->_endpoint, strLength, endpoint);
}

bool UServer::IsStarted(void) const{
	return this->_isStarted;
}

void UServer::CleanUpXS(void){
	if(this->_sock != NULL){
		static_cast<xs::socket_t*>(this->_sock)->close();
		delete static_cast<xs::socket_t*>(this->_sock);

		this->_sock = NULL;
	}

	if(this->_ctx != NULL){
		delete static_cast<xs::context_t*>(this->_ctx);

		this->_ctx = NULL;
	}
}

UServer::~UServer(void){
	this->_isStarted = false;

	delete[] this->_endpoint;
	this->_endpoint = NULL;

	CleanUpXS();
}
